### ISSUE TITLE
`lgb.train()` uses reference label instead of new label from `.set_label()`

### ISSUE THREAD

**Issue by jrvmalik** [microsoft/LightGBM#5067]
## Description
When calling `lgb.train()` on a dataset that has been constructed with a `.subset()` method, the label used is not the label from the subset but rather the label from the reference dataset.  This is also true for `init_score` and `weights`.

## Reproducible example
```
import lightgbm as lgb
import numpy as np

np.random.seed(0)

X = np.random.rand(3000, 3)
y = np.random.rand(3000)

data = lgb.Dataset(data=X, params={"max_bin": 127})

params = {"objective": "l2", "max_depth": 3, "metric": "mse", "deterministic": True, "seed": 1, "learning_rate": 0.1, "num_iterations": 20, "reg_alpha": 0.2}

trn_indices = np.random.choice(X.shape[0], size=(2000,))
trn_data = data.subset(trn_indices)
trn_data.set_label(y[trn_indices]).construct()

trn_data.get_label()

mdl = lgb.train(params, trn_data, num_boost_round=10)
mdl.predict(X)
```
We can see there is a problem because all the predictions are zero.
However we can fix the problem by setting the label of the reference dataset first, and then calling `subset()`.

## Environment info

LightGBM version or commit hash: 3.2.1.99

Command(s) you used to install LightGBM

```shell
pip install lightgbm
```

**Comment by jmoralez** [comment]
Hi @jrvmalik, thank you for your interest in LightGBM. The `subset` method copies the bins and the slice of the data from the original dataset: https://github.com/microsoft/LightGBM/blob/31ab4d42312a6973b4b8620317ef17b050d7eb50/src/c_api.cpp#L1417-L1420
This happens when you construct it, so calling `set_label` before only sets the Python attribute: https://github.com/microsoft/LightGBM/blob/31ab4d42312a6973b4b8620317ef17b050d7eb50/python-package/lightgbm/basic.py#L2145-L2150

If you switch the order it should work as expected, i.e.
```python
import lightgbm as lgb
import numpy as np

np.random.seed(0)

X = np.random.rand(3000, 3)
y = np.random.rand(3000)

data = lgb.Dataset(data=X, params={"max_bin": 127})

params = {"objective": "l2", "max_depth": 3, "metric": "mse", "deterministic": True, "seed": 1, "learning_rate": 0.1, "num_iterations": 20, "reg_alpha": 0.2}

trn_indices = np.random.choice(X.shape[0], size=(2000,))
trn_data = data.subset(trn_indices)
trn_data.construct()
trn_data.set_label(y[trn_indices])

mdl = lgb.train(params, trn_data, num_boost_round=10)
mdl.predict(X)
```

Please let us know if this helps.

**Comment by jrvmalik** [comment]
I see, thanks.  IMO this behaviour of `set_label()` is not intuitive.  For example, this works:
```
import lightgbm as lgb
import numpy as np

np.random.seed(0)

X = np.random.rand(3000, 3)
y = np.random.rand(3000)

data = lgb.Dataset(data=X, params={"max_bin": 127})
data.set_label(y)
data.construct()
data.get_field('label')
```
What is inconsistent is the Python attribute being recognized during construction.

**Comment by jmoralez** [comment]
Yeah I agree. If you tried to call `data.set_field('label', y)` (which is what gets called if the dataset has been constructed) instead of `data.set_label(y)` you'd get a more informative error: `Exception: Cannot set label before construct dataset`. Might be worth removing that if statement and raise that error if you try to set an attribute before constructing the dataset.

**Comment by shiyu1994** [comment]
Thanks for using LightGBM! I believe this is a bug that we need to fix. After setting label with `set_label`, though the `Dataset` has not been constructed, we should noticed that the label of subset has been reset. And when constructing the subset, we should use the set label instead of the label in the full dataset. WDYT @jmoralez ?

**Comment by jrvmalik** [comment]
Thanks @shiyu1994 and @jmoralez . I think it boils down to the question, Is there a reason why `set_label` and `set_field` have different behaviours?

**Comment by Earthson** [comment]
> 

I have a case: Feature is really large to fully loaded, so it is constructed and saved.
Is it possiable to get a subset with new label without original feature X?
